src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2748,6 +2768,8 @@ StartPlacementExecutionOnSession(TaskPlacementExecution *placementExecution,
 
 	/* connection is going to be in use */
 	workerPool->idleConnectionCount--;
+	session->currentTask = placementExecution;
+	placementExecution->executionState = PLACEMENT_EXECUTION_RUNNING;
```

Comment:
Adding test to this PR is extremely hard. Even manually triggering the situation is very hard -- I wasn't able to do yet.

Reply:
I mean, I can only do it with code changes:
```C
index ae3f23ae0..560f5b8e6 100644
--- a/src/backend/distributed/executor/adaptive_executor.c
+++ b/src/backend/distributed/executor/adaptive_executor.c
@@ -1799,7 +1799,7 @@ ManageWorkerPool(WorkerPool *workerPool)
        Assert(initiatedConnectionCount >= activeConnectionCount);
 
        /* we should never have less than 0 connections ever */
-       Assert(activeConnectionCount >= 0 && idleConnectionCount >= 0);
+       //Assert(activeConnectionCount >= 0 && idleConnectionCount >= 0);
 
        if (workerPool->failed)
        {
@@ -2679,6 +2679,7 @@ PopAssignedPlacementExecution(WorkerSession *session)
 }
 
 
+
 /*
  * PopAssignedPlacementExecution finds an executable task from the queue of assigned tasks.
  */
@@ -2749,6 +2750,7 @@ StartPlacementExecutionOnSession(TaskPlacementExecution *placementExecution,
        /* connection is going to be in use */
        workerPool->idleConnectionCount--;
 
+       ShutdownConnection(connection);
        if (paramListInfo != NULL)
        {
                int parameterCount = paramListInfo->numParams;
```